README
************** GLUE 2.0 LDAP implementation - Official Release ****************
  URL:     http://redmine.ogf.org/projects/glue-wg/
           https://github.com/OGF-GLUE/LDAP/
  Authors: David Horat (david.horat@cern.ch), CERN
           Laurence Field (laurence.field@cern.ch), CERN
           Florido Paganelli (florido.paganelli@hep.lu.se), Lund University
***********************************************************i*******************
+++++++++++++++++++++++++++
Version history
+++++++++++++++++++++++++++
v0.1 - Initial release. 25/03/2009
v0.2 - OID and description corrections. Relationship implementation. 30/03/2009
v0.3 - New version with changes agreed by the comunity. 25/05/2009
v2.0rcX - experimental revision carried on by EMI. 10/2012
V2.0 - final version in sync with LDAP realisation document

+++++++++++++++++++++++++++
  Synopsis
+++++++++++++++++++++++++++
  GLUE 2.0 LDAP implementation is an LDAP implementation of the GLUE 2.0 grid
interoperability specifications which can be found at [1]. This package intends
to cover all schemas needed and ldif test files for each object. Moreover, it
will include the necessary scripts to automate both its usage and tests.

+++++++++++++++++++++++++++
File description
+++++++++++++++++++++++++++
Files included in the package:
ldif/			Directory where all LDIF files for testing are stored
  00-Setup.ldif 	LDIF file that setups the environment for all testing
  *.ldif		LDIF files for each object in the specification
  99-TearDown.td	DN of all objects created in the tests
schema/			Directory containing all the LDAP schemas and a version file.
  00-Version.schema     Dummy schema file containing only comments and a
                        version number.
                        The version number should always be in sync with
                        the major version number in this document.
  *.schema		LDAP schemas for development purposes
compile-schema.sh	Script to compile all schemas into a single file
README			This file
takeOffGluePrefix.sh	Script used to take off the 'Glue' prefix from all files
test-ldif.sh		Script to test all LDIF files
ChangeLog.txt   File that contains the ChangeLog of each version of the schema.

Files created when executing scripts:
GLUE20.schema		The file created by compile-schema.sh
GLUE20.schema.bak	Backup copy of the last version of the schemas
test-ldif.log		Log of the test-ldif.sh script

+++++++++++++++++++++++++++
Getting started
+++++++++++++++++++++++++++
This package is meant to be tested in an empty LDAP server with a database
called 'o=grid'.

1) compile-schema.sh
Change directory to where this file is located and execute it. It will output
a file called 'GLUE20.schema' which is a compilation of all schemas under the
schema/ directory. If the file 'GLUE20.schema' already exists, it will copy it
into 'GLUE20.schema.bak'.

2) Schema debugging example with OpenLDAP and Ubuntu
# Create a dir to hold the schemas
sudo mkdir /etc/ldap/schema/glue/
# Copy the schema files to your schemas directory
sudo cp *.schema /etc/ldap/schema/glue/
# Edit your OpenLDAP config file and add lines like
# include /etc/ldap/schema/glue/5-01-Entity.schema
sudo nano /etc/ldap/ldap.conf
# Stop any OpenLDAP running on your system
sudo killall slapd
# Execute OpenLDAP using that config file
sudo slapd -f /etc/ldap/ldap.conf
# View the logs
tail -n 20 /var/log/syslog

3) test-ldif.sh
It will add all LDIF files under the ldif/ directory into an LDAP server
in localhost with the default password 'secret'. Then it will delete all
the tests putting the server in its original state. All logs are kept into
the log file 'test-ldif.log'.

4) Testing single files
- You should always load first the setup LDIF file:
ldapadd -x -D "o=grid" -w secret -h localhost -v -f 00-Setup.ldif
- Then you can test any file, for example "5-03-Location.ldif":
ldapadd -x -D "o=grid" -w secret -h localhost -v -f 5-03-Location.ldif
- After, you should run the tear down file:
ldapdelete -x -D "o=grid" -w secret -h localhost -v -f 99-TearDown.td

NOTE: If the default values do not fit your needs, feel free to edit the scripts
and change them. It is easy to do! :)



ChangeLog.txt
########################################################################
#
# Changelog for LDAP GLUE2 Schema
#
########################################################################

v2.0 Florido Paganelli 2015-05-12
   * Changed URLs in documents to reflect current status

v2.0 Florido Paganelli 2014-07-02

    * Changed release to 2.0
    
v2.0rc7 Florido Paganelli 2014-05-05

   * Changed Domain, Resource, Share, Policy to STRUCTURAL to be 
     consistent with realisation document.   

v2.0rc6 Florido Paganelli 2014-05-05

   * Changed Domain, Resource, Share, Policy to ABSTRACT to be 
     consistent with model and XML
   * Changed ALL other objectclasses to STRUCTURAL
   * Changed all Boolean fields to DirectoryString, that is, all
     SYNTAX 1.3.6.1.4.1.1466.115.121.1.7
     to
     SYNTAX 1.3.6.1.4.1.1466.115.121.1.15
     NOTE: this will remove boolean attributes without values from the 
     rendering.
   
v2.0rc5 Florido Paganelli 2013-01-11
  
   * Changed GLUE2GroupName attribute NAME back to GLUE2GroupID, added GLUE2GroupName attribute as an alias
   * Changed GLUE2Group object MUST back to GLUE2GroupID
   * This solution preserves backward compatibility between EMI3 and EMI2 BDII products, that was broken with 2.0rc4.

v2.0rc4 Florido Paganelli 2012-11-01
  
   * Changed GLUE2GroupID attribute to GLUE2GroupName, added attribute aliases
   * Changed MUST to GLUE2GroupName
   * The two solutions above preserve full compatibility.
   
v2.0rc3 Florido Paganelli 2012-10-14
  
   * Changed Type to DirectoryString (SYNTAX 1.3.6.1.4.1.1466.115.121.1.15)
     on all String attributetypes to align with LDAP Rendering DRAFT.
     Changed EQUALITY and SUBSTR to be compatible with DirectoryString.
   !!ATTENTION!!: there are issues with DirectoryString. A DirectoryString
                attribute cannot be empty. Maybe this rc3 must be discarded.

v2.0rc2 Florido Paganelli 2012-10-12
  
   * Changed OIDs on all objects.
     A summary of the new OIDs numbering can be found at
       https://twiki.cern.ch/twiki/pub/EMI/Glue2LdapStructure/ReassigningGLUE2OIDs.pdf

v2.0rc1 Florido Paganelli 2012-10-10

  * Added Version file
  * changed all relevant entities from AUXILIARY to STRUCTURAL:
    
    Changed AUXILIARY to
    
    SUP GLUE2<Superclass>
    STRUCTURAL
    
    in AdminDomain,UserDomain,
    ComputingService, ComputingEndpoint, ComputingShare, ComputingManager,
    ExecutionEnvironment, StorageService,
    StorageEndpoint, StorageShare


compile-schema.sh
#!/bin/bash

################################################################################
#                  GLUE 2.0 LDAP schema compilation script
#
# URL:       http://forge.gridforum.org/sf/projects/glue-wg
# Author:    David Horat (david.horat@cern.ch), CERN
# Version:   0.2
# Changelog: 0.1 (24/03/2009) - First version
#            0.2 (13/07/2009) - Added copyright notice
################################################################################

# Basic info
`unalias -a` # Remove all aliases from this new shell
STARTTIME=`date +%s`
VERSION='0.2'
COPYRIGHT='LICENSE'

# Parameters (Adapt them to your needs)
INPUTDIR='schema'
OUTPUTFILE='GLUE20.schema'

# Welcome message
echo "Welcome to GLUE 2.0 LDAP schema compilation script v$VERSION"
date

# Check if needed files and dirs are present
if [ ! -d $INPUTDIR ]; then
  echo "Input dir '$INPUTDIR' does not exist."
  exit 1
fi

# If OUTPUTFILE exist, make a backup and delete it
if [ -f $OUTPUTFILE ]; then
  mv -f $OUTPUTFILE $OUTPUTFILE.'bak'
fi

# Add the copyright notice to the new file
cat $COPYRIGHT > $OUTPUTFILE

# Collect all content of the files from the INPUTDIR and put them into OUTPUTFILE
for file in $INPUTDIR/*.schema; do 
  echo '# File:   ' $file >> $OUTPUTFILE
  cat $file >> $OUTPUTFILE
  echo '' >> $OUTPUTFILE
done

# Calculate runtime and bye messages
RUNTIME=$[`date +%s` - $STARTTIME]
echo "This script took $RUNTIME seconds to run."
echo "$OUTPUTFILE file is ready!"


test-ldif.sh
#!/bin/bash

################################################################################
#                  GLUE 2.0 LDAP unit testing script
#
# URL:     http://forge.gridforum.org/sf/projects/glue-wg
# Author:  David Horat (david.horat@cern.ch), CERN
# Version: 0.1 (24/03/2009)
# Version: 0.2 (13/10/2010) Changed localhost for `echo $HOSTNAME`
################################################################################

# Basic info
`unalias -a` # Remove all aliases from this new shell
STARTTIME=`date +%s`
VERSION='0.1'

# Parameters (Adapt them to your needs)
INPUTDIR='ldif'
TEARDOWN="$INPUTDIR/99-TearDown.td"
HOST=`echo $HOSTNAME`
PASS='bdiidbpassword' # This should be used just in a testing machine
LOGFILE='test-ldif.log'

# Welcome message, erase the old log file
echo "Welcome to GLUE 2.0 LDAP unit testing script v$VERSION"
date
rm -f $LOGFILE

# Check if needed files and dirs are present
if [ ! -d $INPUTDIR ]; then
  echo "Input dir '$INPUTDIR' does not exist."
  exit 1
fi
if [ ! -f $TEARDOWN ]; then
  echo "Tear down file '$INPUTDIR' does not exist."
  exit 1
fi

# Collect all content of the files from the INPUTDIR and put them into OUTPUTFILE
for file in $INPUTDIR/*.ldif; do
  echo "Testing file $file"
  ldapadd -x -D "o=glue" -w $PASS -h $HOST -v -f $file &> $LOGFILE
  if [ $? -ne 0 ]; then
    echo "Errors when trying to add the file $file"
    RUNTIME=$[`date +%s` - $STARTTIME]
    echo "This script took $RUNTIME second/s to run."
    echo "ERROR! Check the log in $LOGFILE"
    exit 2
  fi
done

# Tear down to recover initial state
echo 'Tearing down'
ldapdelete -x -D "o=glue" -w $PASS -h $HOST -v -f $TEARDOWN &> $LOGFILE
if [ $? -ne 0 ]; then
  echo "Errors when in the tear down process using the file $TEARDOWN"
  RUNTIME=$[`date +%s` - $STARTTIME]
  echo "This script took $RUNTIME second/s to run."
  echo "ERROR! Check the log in $LOGFILE"
  exit 3
fi

# Calculate runtime and bye messages
RUNTIME=$[`date +%s` - $STARTTIME]
echo "This script took $RUNTIME second/s to run."
echo 'All OK!'
